stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -490,10 +497,106 @@ impl NakamotoBlockHeader {
     }
 
     /// Verify the block header against an aggregate public key
-    pub fn verify_signer(&self, signer_aggregate: &Point) -> bool {
-        let schnorr_signature = &self.signer_signature.0;
+    pub fn verify_threshold_signer(
+        &self,
+        signer_aggregate: &Point,
+        signature: &ThresholdSignature,
+    ) -> bool {
         let message = self.signer_signature_hash().0;
-        schnorr_signature.verify(signer_aggregate, &message)
+        signature.verify(signer_aggregate, &message)
+    }
+
+    /// Verify the block header against the list of signer signatures
+    ///
+    /// Validate against:
+    /// - Any invalid signatures (eg not recoverable or not from a signer)
+    /// - Any duplicate signatures
+    /// - At least the minimum number of signatures (based on total signer weight
+    /// and a 70% threshold)
+    /// - Order of signatures is maintained vs signer set
+    pub fn verify_signer_signatures(&self, reward_set: &RewardSet) -> Result<(), ChainstateError> {
+        let message = self.signer_signature_hash();
+        let Some(signers) = &reward_set.signers else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "No signers in the reward set".to_string(),
+            ));
+        };
+
+        let mut total_weight_signed: u32 = 0;
+        // `last_index` is used to prevent out-of-order signatures
+        let mut last_index = None;
+
+        let total_weight = signers.iter().map(|s| s.weight).fold(0, |w, acc| {
+            acc.checked_add(w)
+                .expect("FATAL: Total signer weight > u32::MAX")
+        });
+
+        // HashMap of <PublicKey, (Signer, Index)>
+        let signers_by_pk = signers
```

Comment:
Nit: can you avoid the superfish operator and instead use `let signers_by_pubkey : HashMap<_, _> = /* ... */`?